### PR TITLE
Add cross api version breaking change detector

### PR DESCRIPTION
**Pull request by jianyexi**
Also refactor some lintDiff code

### CODE REVIEW

**Comment by jianyexi** [comment]
> _No description provided._

File: src/breaking-change.ts [hunk 1]
```diff
@@ -162,13 +165,183 @@ async function runOad(oldSpec: string, newSpec: string) {
...
+    const folder = this.getRPFolder(swaggerFile);
+    if (!folder) {
+      return swaggerMetaData
+    }
+    const allSwaggers = this.getAllSwaggers(folder);
+    for (const swagger of allSwaggers) {
+      const version = getVersionFromInputFile(swagger);
+      const fileName = path.basename(swagger);
+      // Needs to consider other cases
+      const versionType = swagger.includes("preview") ? "preview" : "stable";
```
Comment (tiramisuq) [inline src/breaking-change.ts:225]: Should apiVersion be examined with 'privatepreview'

Comment (jianyexi) [reply src/breaking-change.ts:225]: fixed

File: src/breaking-change.ts [hunk 2]
```diff
@@ -162,13 +165,183 @@ async function runOad(oldSpec: string, newSpec: string) {
...
+  pr : PullRequestProperties
+  versionManager: SwaggerVersionManager = new SwaggerVersionManager()
+  constructor(pullRequest: PullRequestProperties,newSwaggers:string[]) {
+    this.swaggers = newSwaggers
+    this.pr = pullRequest
+  }
+
+  async diffOne(oldSpec: string ,newSpec : string) {
+    try {
+      await runOad(path.resolve(this.pr!.workingDir,oldSpec ), newSpec);
```
Comment (ruowan) [inline src/breaking-change.ts:272]: code format.

File: src/breaking-change.ts [hunk 3]
```diff
@@ -162,13 +165,183 @@ async function runOad(oldSpec: string, newSpec: string) {
...
+      const previous = await utils.doOnTargetBranch(this.pr, async () => {
+        return this.versionManager.getClosestPreview(swagger)
+      })
+      if (previous) {
+        await this.diffOne(path.resolve(this.pr!.workingDir,previous), swagger);
+      }
+    }
+  }
+
+  async getBreakingChangeBaseOnStableVersion() {
```
Comment (ruowan) [inline src/breaking-change.ts:300]: naming. get => detect.  get means return something.

File: src/breaking-change.ts [hunk 4]
```diff
@@ -197,16 +383,9 @@ export async function runScript() {
   console.log("Processing swaggers:");
   console.log(swaggersToProcess);
 
-  console.log("Finding new swaggers...");
+  changeTargetBranch(pr)
```
Comment (ruowan) [inline src/breaking-change.ts:386]: Why changeTargetBranch

Comment (jianyexi) [reply src/breaking-change.ts:386]: Against master

File: src/breaking-change.ts (hunk 2 above)
Comment (jianyexi) [reply src/breaking-change.ts]: fixed

File: src/breaking-change.ts (hunk 3 above)
Comment (jianyexi) [reply src/breaking-change.ts]: Oh, thanks , fixed it

File: src/breaking-change.ts [hunk 5]
```diff
@@ -162,13 +165,183 @@ async function runOad(oldSpec: string, newSpec: string) {
...
+      new: it.new,
+      old: it.old,
+      details: utils.cutoffMsg(it.error.stack) || "",
+    },
+  }));
+
+  fs.appendFileSync("pipe.log", JSON.stringify(errorResult) + "\n");
+  console.log(`oad error log: ${JSON.stringify(errorResult)}`);
+}
+export class SwaggerVersionManager {
```
[the remark this answers is not in the document]
Comment (jianyexi) [reply src/breaking-change.ts:197]: Added

**Review by tiramisuq — APPROVED** [review]

File: src/breaking-change.ts [hunk 6]
```diff
@@ -8,6 +8,9 @@ import * as path from "path";
 import * as tsUtils from "./ts-utils";
 import { targetHref } from "./utils";
 import * as utils from "./utils";
+import { PullRequestProperties } from '@azure/avocado/dist/dev-ops';
```
[the remark this answers is not in the document]
Comment (jianyexi) [reply src/breaking-change.ts]: removed

File: src/breaking-change.ts (hunk 4 above)
Comment (jianyexi) [reply src/breaking-change.ts:396]: Move old code to the function

**Review by ruowan — APPROVED** [review]

**Pull request merged**